### ISSUE TITLE
Position 2 is out of bound in Block::safeGetByPosition()

### ISSUE THREAD

**Issue by skyoct** [ByConity/ByConity#1757]
### Briefly describe the bug
```
select ck_date, packagename, count(distinct packagename) as ck_result
from ods.app_10001_mc_default
where ck_date in ('2024-07-16')
  and toTimeZone(ck_datetime_timestamp, 'Asia/Shanghai') >= '2024-07-16 00:00:00'
  and toTimeZone(ck_datetime_timestamp, 'Asia/Shanghai') <= '2024-07-16 23:59:59'
  and pd_kind = 'event_ad_intercept'
  and `adplaytype` = 'load_call_reward'
group by ck_date, packagename
order by ck_date, ck_result desc
limit 50000
settings
max_execution_time = 0, enable_optimizer=1;
```

An error occurred when running the above SQL statement. When the optimizer is turned off, the query runs normally.

```
2024.07.17 16:12:51.218014 [ 22743 ] {26623a19-f826-43d9-8cb6-aa48251c6696_1} (451203293669752833) <Error> PlanSegmentExecutor: [26623a19-f826-43d9-8cb6-aa48251c6696_1]: Query has excpetion with code: 11, detail
 
: Code: 11, e.displayText() = DB::Exception: Position 2 is out of bound in Block::safeGetByPosition(), max position = 1, there are columns: ck_date, packagename: While executing MergingAggregatedTransform SQLSTA
TE: HY000, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x274670f2 in /opt/byconity/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x10946060 in /opt/byconity/bin/clickhouse
2. DB::Block::safeGetByPosition(unsigned long) @ 0x2058e5ea in /opt/byconity/bin/clickhouse
3. void DB::Aggregator::mergeStreamsImplCase<false, DB::AggregationMethodSerialized<HashMapTable<StringRef, HashMapCellWithSavedHash<StringRef, char*, DefaultHash<StringRef, void>, HashTableNoState>, DefaultHash
<StringRef, void>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> > >, HashMapTable<StringRef, HashMapCellWithSavedHash<StringRef, char*, DefaultHash<StringRef, void>, HashTableNoState>, DefaultHa
sh<StringRef, void>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> > >(DB::Block&, DB::Arena*, DB::AggregationMethodSerialized<HashMapTable<StringRef, HashMapCellWithSavedHash<StringRef, char*, D
efaultHash<StringRef, void>, HashTableNoState>, DefaultHash<StringRef, void>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> > >&, HashMapTable<StringRef, HashMapCellWithSavedHash<StringRef, char*
, DefaultHash<StringRef, void>, HashTableNoState>, DefaultHash<StringRef, void>, HashTableGrowerWithPrecalculation<8ul>, Allocator<true, true> >&, char*) const @ 0x2138272b in /opt/byconity/bin/clickhouse
4. DB::Aggregator::mergeBlocks(std::__1::map<int, std::__1::list<DB::Block, std::__1::allocator<DB::Block> >, std::__1::less<int>, std::__1::allocator<std::__1::pair<int const, std::__1::list<DB::Block, std::__1
::allocator<DB::Block> > > > >, DB::AggregatedDataVariants&, unsigned long) @ 0x21241bc5 in /opt/byconity/bin/clickhouse
5. DB::MergingAggregatedTransform::generate() @ 0x228485e3 in /opt/byconity/bin/clickhouse
6. DB::IAccumulatingTransform::work() @ 0x224d1e70 in /opt/byconity/bin/clickhouse
7. void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::$_0, void ()> >(std::__1::__function::__po
licy_storage const*) @ 0x224e87c6 in /opt/byconity/bin/clickhouse
8. DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) @ 0x224e6997 in /opt/byconity/bin/clickhouse
9. void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PipelineExecutor::executeImpl(unsigned long)::$_5>(D
B::PipelineExecutor::executeImpl(unsigned long)::$_5&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0x224e9405 in /opt/byconity/bin/clickhouse
10. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x10982840 in /opt/byconity/bin/clickhouse
11. void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>
(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) @ 0x10986c3a in /opt/byconity/bin/clickhouse
12. start_thread @ 0x7ea7 in /lib/x86_64-linux-gnu/libpthread-2.31.so
13. __clone @ 0xfca2f in /lib/x86_64-linux-gnu/libc-2.31.so
 (version 21.8.7.1)
```

### Version
Build between 0.4.1 and 0.4.2

**Comment by superhail** [comment]
can u provide a minimum reproducable example?

**Comment by superhail** [comment]
try with enable_single_distinct_to_group_by=0

**Comment by superhail** [comment]
@nudles this one needs to be merged to Byconity, meego ID:4086580387

**Comment by nudles** [comment]
> @nudles this one needs to be merged to Byconity, meego ID:4086580387

merged into byconity 1.0
